File: java/InvertBinaryTree226.java
/*
Given the root of a binary tree, invert the tree, and return its root.
*/
public class InvertBinaryTree226 {
    public static void main(String[] args) {
        TreeNode root = new TreeNode(4, new TreeNode(2, new TreeNode(1), new TreeNode(3)), new TreeNode(7, new TreeNode(6), new TreeNode(9)));
        TreeNode res = invertTree(root);
        System.out.println(res.val);

        root = new TreeNode(2, new TreeNode(1), new TreeNode(3));
        res = invertTree(root);
        System.out.println(res.val);

        root = new TreeNode(1);
        res = invertTree(root);
        System.out.println(res.val);
    }

    public static TreeNode invertTree(TreeNode root) {
        if (root == null) return null;
        TreeNode leftTemp = invertTree(root.right);
        root.right = invertTree(root.left);
        root.left = leftTemp;
        return root;
    }

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

File: python/InvertBinaryTree226.py
# Given the root of a binary tree, invert the tree, and return its root.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root:
        return
    root.left, root.right = invert_tree(root.right), invert_tree(root.left)
    return root

    # if root:
    #     stack = [root]
    #     while stack:
    #         node = stack.pop()
    #         if node:
    #             node.left, node.right = node.right, node.left
    #             stack += node.left, node.right
    # return root


def print_tree(root: Optional[TreeNode], indent: str = ''):
    if not root:
        return
    print_tree(root.right, indent + '   ')
    print(indent, root.val)
    print_tree(root.left, indent + '   ')


if __name__ == "__main__":
    tree_root = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(7, TreeNode(6), TreeNode(9)))
    print_tree(tree_root)
    print_tree(invert_tree(tree_root))
